app/controllers/interest_controller.rb

```diff
@@ -25,100 +25,133 @@ class InterestController < ApplicationController
+    notifications = find_relevant_notifications
+    interests = find_relevant_interests
+    @targets = notifications + interests
+
+    @pages = paginate_numbers(:page, 50)
+    @pages.num_total = @targets.length
+    @targets = @targets[@pages.from..@pages.to]
+  end
+
+  private
+
+  def find_relevant_notifications
+    Notification.for_user(@user).sort do |a, b|
       result = a.flavor.to_s <=> b.flavor.to_s
       result = a.summary.to_s <=> b.summary.to_s if result.zero?
       result
     end
-    # interests = Interest.find_all_by_user_id(@user.id).sort do |a,b|
-    interests = Interest.where(user_id: @user.id).sort do |a, b|
+  end
+
+  def find_relevant_interests
+    Interest.for_user(@user).sort do |a, b|
       result = a.target_type <=> b.target_type
       if result.zero?
         result = (a.target ? a.target.text_name : "") <=>
                  (b.target ? b.target.text_name : "")
       end
       result
     end
-    @targets = notifications + interests
-    @pages = paginate_numbers(:page, 50)
-    @pages.num_total = @targets.length
-    @targets = @targets[@pages.from..@pages.to]
   end
 
+  public
+
   # Callback to change interest state in an object.
   # Linked from: show_<object> and emails
   # Redirects back (falls back on show_<object>)
   # Inputs: params[:type], params[:id], params[:state], params[:user]
   # Outputs: none
-  def set_interest
+  def set_interest # rubocop:disable Metrics/AbcSize
     pass_query_params
-    type   = params[:type].to_s
-    oid    = params[:id].to_i
-    state  = params[:state].to_i
-    uid    = params[:user]
-    target = Comment.find_object(type, oid)
-    if @user
-      interest = Interest.find_by(
-        target_type: type, target_id: oid, user_id: @user.id
+    target_type = params[:type].to_s
+    target_id   = params[:id].to_i
+    state       = params[:state].to_i
+    user_id     = params[:user]
+    target      = Comment.find_object(target_type, target_id)
+    return unless @user
+
+    interest = Interest.find_by(
+      target_type: target_type, target_id: target_id, user_id: @user.id
+    )
+    if user_id && @user.id != user_id.to_i
+      flash_error(:set_interest_user_mismatch.l)
+    elsif !target && state != 0
+      flash_error(:set_interest_bad_object.l(type: target_type, id: target_id))
+    else
+      set_interest_state_for_target(interest, state, target)
+    end
+    redirect_to_target_or_interests(target, target_id)
+  end
+
+  private
+
+  def redirect_to_target_or_interests(target, target_id)
+    unless target
+      return redirect_back_or_default(controller: "/interest",
+                                      action: "list_interests")
+    end
+
+    redirect_back_or_default(
+      add_query_param(controller: target.show_controller,
+                      action: target.show_action, id: target_id)
+    )
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def set_interest_state_for_target(interest, state, target)
+    if !interest && state != 0
```

Comment:
I wonder if this would pass ABC if this initialization of new interests were outsourced?  Once a method is reduced to a simple switch statement, and it still fails ABC, then fair enough.  But maybe this one can actually be fixed without silly contortions.